### ISSUE TITLE
Unnecessary Motion when join on distribution keys of partition tables with others.

### ISSUE THREAD

**Issue by avamingli** [greenplum-db/gpdb#15774]
### Greenplum version or build

7x, may need back-port

### Expected behavior

No motion if tables' locus are compatible for join.

ORCA:
```sql
EXPLAIN(COSTS OFF) SELECT COUNT(*) FROM t1 JOIN t2 using(id);
                            QUERY PLAN
------------------------------------------------------------------
 Aggregate
   ->  Gather Motion 3:1  (slice1; segments: 3)
         ->  Hash Join
               Hash Cond: ((t1_1_prt_1.id)::text = (t2.id)::text)
               ->  Append
                     ->  Seq Scan on t1_1_prt_1
               ->  Hash
                     ->  Seq Scan on t2
 Optimizer: Pivotal Optimizer (GPORCA)
(9 rows)
```

### Step to reproduce the behavior

```sql

CREATE TABLE t1 (id varchar(32), date date) DISTRIBUTED BY (id)
PARTITION BY RANGE (date)
(START (date '2016-01-01') INCLUSIVE END (date '2016-01-02') EXCLUSIVE EVERY (INTERVAL '1 day'));
CREATE TABLE t2 (id varchar(32)) DISTRIBUTED BY (id);
analyze t1;
analyze t2;
\d+ t1;
                                      Partitioned table "public.t1"
 Column |         Type          | Collation | Nullable | Default | Storage  | Stats target | Description 
--------+-----------------------+-----------+----------+---------+----------+--------------+-------------
 id     | character varying(32) |           |          |         | extended |              | 
 date   | date                  |           |          |         | plain    |              | 
Partition key: RANGE (date)
Partitions: t1_1_prt_1 FOR VALUES FROM ('01-01-2016') TO ('01-02-2016')
Distributed by: (id)

\d+ t2;
                                            Table "public.t2"
 Column |         Type          | Collation | Nullable | Default | Storage  | Stats target | Description 
--------+-----------------------+-----------+----------+---------+----------+--------------+-------------
 id     | character varying(32) |           |          |         | extended |              | 
Distributed by: (id)

EXPLAIN(COSTS OFF) SELECT COUNT(*) FROM t1 JOIN t2 USING(id);
                            QUERY PLAN                            
------------------------------------------------------------------
 Aggregate
   ->  Gather Motion 3:1  (slice1; segments: 3)
         ->  Hash Join
               Hash Cond: ((t1.id)::text = (t2.id)::text)
               ->  Redistribute Motion 3:3  (slice2; segments: 3)
                     Hash Key: t1.id
                     ->  Seq Scan on t1_1_prt_1 t1
               ->  Hash
                     ->  Seq Scan on t2
 Optimizer: Postgres query optimizer
(10 rows)
```
Direct join a sub partition table is ok.
```sql
gpadmin=# EXPLAIN(COSTS OFF) SELECT COUNT(*) FROM t1_1_prt_1  JOIN t2 USING(id);
                            QUERY PLAN
------------------------------------------------------------------
 Aggregate
   ->  Gather Motion 3:1  (slice1; segments: 3)
         ->  Hash Join
               Hash Cond: ((t1_1_prt_1.id)::text = (t2.id)::text)
               ->  Seq Scan on t1_1_prt_1
               ->  Hash
                     ->  Seq Scan on t2
 Optimizer: Postgres query optimizer
(8 rows)
```
Have a fix on it, will new a pr soon.

**Comment by kainwen** [comment]
I think this is duplicated with https://github.com/greenplum-db/gpdb/issues/13402
and previously we have a fix https://github.com/greenplum-db/gpdb/pull/13423
but have some concern so closing that PR.

Cherry pick the closed PR can also fix this issue.

We should come back when having time.

Close this issue because we have other issue to mark.

**Comment by avamingli** [comment]
> I think this is duplicated with #13402 and previously we have a fix #13423 but have some concern so closing that PR.
> 
> Cherry pick the closed PR can also fix this issue.
> 
> We should come back when having time.
> 
> Close this issue because we have other issue to mark.

Oh, I didn't search it first. Thanks,  I have the same fix with https://github.com/greenplum-db/gpdb/pull/13423.
